src/macros.rs

```diff
@@ -18,10 +18,83 @@ macro_rules! impl_pretty_debug {
+        #[cfg(feature = "alloc")]
+        use alloc::string::String;
+
+        #[cfg(feature = "bitcoin_hashes")]
+        impl ::core::fmt::Debug for $thing {
+            fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
+                const DEBUG_HASH_TAG: &[u8] = &[
+                    0x66, 0xa6, 0x77, 0x1b, 0x9b, 0x6d, 0xae, 0xa1, 0xb2, 0xee, 0x4e, 0x07, 0x49,
+                    0x4a, 0xac, 0x87, 0xa9, 0xb8, 0x5b, 0x4b, 0x35, 0x02, 0xaa, 0x6d, 0x0f, 0x79,
+                    0xcb, 0x63, 0xe6, 0xf8, 0x66, 0x22
+                ]; // =SHA256(b"rust-secp256k1DEBUG");
+                use ::bitcoin_hashes::{Hash, sha256, HashEngine};
+
+                let mut engine = sha256::HashEngine::default();
+                engine.input(DEBUG_HASH_TAG);
+                engine.input(DEBUG_HASH_TAG);
+                engine.input(&self.0[..]);
+                let hash = sha256::Hash::from_engine(engine);
+
+                write!(f, "{}(#", stringify!($thing))?;
+                for i in &hash[..4] {
+                    write!(f, "{:02x}", i)?;
+                }
+                f.write_str("...")?;
+                for i in &hash[28..] {
+                    write!(f, "{:02x}", i)?;
+                }
+                f.write_str(")")
+            }
+        }
+
+        #[cfg(all(not(feature = "bitcoin_hashes"), feature = "std"))]
+        impl ::core::fmt::Debug for $thing {
+            fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
+                use ::core::hash::Hasher;
+                const DEBUG_HASH_TAG: &[u8] = &[
+                    0x66, 0xa6, 0x77, 0x1b, 0x9b, 0x6d, 0xae, 0xa1, 0xb2, 0xee, 0x4e, 0x07, 0x49,
+                    0x4a, 0xac, 0x87, 0xa9, 0xb8, 0x5b, 0x4b, 0x35, 0x02, 0xaa, 0x6d, 0x0f, 0x79,
+                    0xcb, 0x63, 0xe6, 0xf8, 0x66, 0x22
+                ]; // =SHA256(b"rust-secp256k1DEBUG");
+                let mut hasher = ::std::collections::hash_map::DefaultHasher::new();
+
+                hasher.write(DEBUG_HASH_TAG);
+                hasher.write(DEBUG_HASH_TAG);
+                hasher.write(&self.0[..]);
+                let hash = hasher.finish();
+
+                write!(f, "{}(#{:016x})", stringify!($thing), hash)
+            }
+        }
+
+        impl $thing {
+            /// Formats the explicit byte value of the secret key kept inside the type as a
+            /// little-endian hexadecimal string using the provided formatter.
+            ///
+            /// This is the only method that outputs the actual secret key value, and, thus,
+            /// should be used with extreme precaution.
+            #[deprecated(
+                note = "Caution: you are explicitly outputting secret key value! This can be done
+                only in debug environment and that's why always considered as ``deprecated''"
+            )]
+            pub fn fmt_secret(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
```

Comment:
This function isn't used anywhere

Reply:
But it may be used for printing by library users when a formatter is available (for instance in other Debug/Display implementations)

Reply:
Potentially, a design analogous to [`Path::display`](https://doc.rust-lang.org/std/path/struct.Display.html) from the standard library would be more convenient. It is already possible to get the bytes from a secret key and users can format them whichever way they wish.

You can't actually instantiate a `Formatter` yourself, so this function is only useful if used from within a context of other `Display` and `Debug` impls. If you are already writing your own `Display`/`Debug` for another type, it is not too much to ask to write this trivial loop as well.
However, if what I want is to display the secret key, having to go through the effort of making a newtype is quite burdensome.

Thus, if we want to provide a convenience function for displaying a secret key to hex, I believe it should be as convenient as possible (while still making clear that secret values are being exposed here).

Something like `pub fn display_secret(&self) -> DisplaySecret<'_>` would allow that.